jupyterlab_gcedetails/src/service/notebook_instance_service_layer.ts

```diff
@@ -0,0 +1,453 @@
+   * server if not set.
+   */
+  private async _getProject(): Promise<string> {
+    if (this.projectIdPromise) {
+      return this.projectIdPromise;
+    }
+    try {
+      const { project } = await this._getMetadata();
+      this.projectIdPromise = Promise.resolve(project);
+      return this.projectIdPromise;
+    } catch (err) {
+      console.error('Unable to obtain GCP Project', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Returns the instanceName from the object or tries to retrieve it from the
+   * server if not set.
+   */
+  private async _getInstance(): Promise<string> {
+    if (this.instanceNamePromise) {
+      return this.instanceNamePromise;
+    }
+    try {
+      const { name } = await this._getMetadata();
+      this.instanceNamePromise = Promise.resolve(name);
+      return this.instanceNamePromise;
+    } catch (err) {
+      console.error('Unable to obtain GCP Instance Name', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Returns the locationId from the object or tries to retrieve it from the
+   * server if not set.
+   */
+  private async _getLocation(): Promise<string> {
+    if (this.locationIdPromise) {
+      return this.locationIdPromise;
+    }
+    try {
+      const { zone } = await this._getMetadata();
+      // extract the location from the zone resource name
+      const locationId = zone.split('/')[3];
+      this.locationIdPromise = Promise.resolve(locationId);
+      return this.locationIdPromise;
+    } catch (err) {
+      console.error('Unable to obtain GCP Zone', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Retrieves the VM Metadata from the server
+   */
+  private _getMetadata(): Promise<InstanceMetadata> {
+    if (this.metadataPromise) {
+      return this.metadataPromise;
+    }
+    this.metadataPromise = getMetadata().catch(err => {
+      console.error('Unable to obtain VM Metadata');
+      throw err;
+    });
+    return this.metadataPromise;
+  }
+
+  /**
+   * Stops the notebook instance.
+   */
+  async stop(): Promise<Instance> {
+    try {
+      const [projectId, instanceName, locationId] = await Promise.all([
+        this.projectId,
+        this.instanceName,
+        this.locationId,
+      ]);
+      const name = `projects/${projectId}/locations/${locationId}/instances/${instanceName}`;
+      const pendingOperation = await this._transportService.submit<Operation>({
+        path: `${NOTEBOOKS_API_PATH}/${name}:stop`,
+        method: POST,
+        headers: { Authorization: `Bearer ${this._authToken}` },
+        body: {},
+      });
+      const finishedOperation = await this._pollOperation(
+        `${NOTEBOOKS_API_PATH}/${pendingOperation.result.name}`
+      );
+      return finishedOperation.response as Instance;
+    } catch (err) {
+      console.error('Unable to stop notebook instance');
+      handleApiError(err);
+    }
+  }
+
+  /**
+   * Starts the notebook instance.
+   */
+  async start(): Promise<Instance> {
```

Comment:
The code for this method is virtually identically to the code for `stop`. I'd consider having each method call a private method with a parameter typed as `start|stop` to affect the `path` value since that's the only difference.

Reply:
Since the setAccelerator and setMachineType are also very similar apart from the request path, request method, request body and error, would it be good practice to have these methods use the private method as well or would that be too much? I could then just pass an error message and request object to the private buildRequest method and it may reduce duplicated code a bit more

Reply:
I'd probably say it's worth factoring those two methods to use some common method as well since each of them return a long-running operation that eventually resolves to an `Instance`.